File: Week 1/python/14.how_other_datatypes_are_stored.py
"""
WEEK 1 - PYTHON FUNDAMENTALS
Topic: How Float and Character Data Are Stored
File: 14.how_other_datatypes_are_stored.py

CONCEPT:
Python's `float` is always a 64-bit IEEE 754 double — there is NO separate
single-precision float by default (numpy provides one). Python has NO `char`
type; characters are length-1 `str` values whose Unicode code point is
returned by `ord()` and the inverse is `chr()`.

KEY POINTS:
- float is 8 bytes (IEEE 754 double); use `struct.pack('!f', x)` for 32-bit form
- No 'f' suffix needed — `1.5` is a float; `1.5e10` is scientific notation
- `Decimal` (decimal module) gives exact base-10 arithmetic when needed
- A character is just `str` of length 1 — `ord('A') == 65`, `chr(65) == 'A'`
- Adding an int to a character requires `chr(ord(ch) + n)` — Python does NOT
  silently coerce char + int to int the way Java does

SYNTAX:
x = 10.4               # float (always double precision)
ord('A')               # 65
chr(65)                # 'A'
chr(ord('a') + 1)      # 'b'

DRY RUN:
1. ord('X') -> 88, ord('Y') -> 89
   print(chr(88), chr(89)) -> "X Y"
2. chr(ord('a') + 1) -> 'b'   (Java prints int 98)
3. ord('a') + ord('b') -> 195  (Java would also produce 195 because chars are ints)

COMPLEXITY: O(1) for ord/chr.
"""

import struct
from decimal import Decimal


def show_float_storage() -> None:
    print("=== Float storage ===")
    x = 10.4
    print(f"x          = {x}")
    print(f"size (sys) = {struct.calcsize('d')} bytes  (double precision)")
    print(f"hex repr   = {x.hex()}")                     # IEEE 754 hex
    # Single-precision form via struct.pack
    single = struct.pack("!f", x)
    print(f"32-bit IEEE = {single.hex()}  (lossy: rounds to nearest float32)")
    print(f"Decimal    = {Decimal('10.4')}  (exact base-10 representation)")


def show_char_storage() -> None:
    print("\n=== Character storage (no `char` type — strings of length 1) ===")
    ch1 = chr(88)            # 'X'  (Java: char ch1 = 88;)
    ch2 = "Y"
    print(f"ch1, ch2 = {ch1}, {ch2}")

    # Adding int to a "char": MUST use ord/chr explicitly in Python
    print(f"chr(ord('a') + 1) = {chr(ord('a') + 1)}     # 'b'")

    # Java prints `'a' + 1` as 98; Python equivalent:
    print(f"ord('a') + 1      = {ord('a') + 1}        # 98 (int)")
    print(f"ord('a') + ord('b') = {ord('a') + ord('b')}      # 195 (Java behaviour)")

    # Concatenation behaviour
    # In Java: 'a' + 'b' (chars) -> 195 (int); ('a'+'b' as Strings) -> "ab"
    # In Python: 'a' + 'b' (strs) -> "ab"
    print(f"'a' + 'b'         = {'a' + 'b'}        # 'ab' (string concat)")


def main() -> None:
    show_float_storage()
    show_char_storage()


if __name__ == "__main__":
    main()


# NOTES:
# - Java needs the 'f' suffix to declare a float literal (`10.4f`); Python has
#   no separate single-precision literal — every `1.5` is a 64-bit double.
# - For exact decimal math (e.g. money), use `decimal.Decimal` in Python.
# - For arrays of fixed-width numerics, use `array.array` or numpy.
# - Characters in Python are NEVER implicitly numeric — `'a' + 1` is a TypeError.

File: Week 1/cpp/14.how_other_datatypes_are_stored.cpp
/*
 * WEEK 1 - C++ FUNDAMENTALS
 * Topic: How float, double, and char are Stored
 * File: 14.how_other_datatypes_are_stored.cpp
 *
 * CONCEPT:
 * C++ distinguishes float (IEEE 754 single precision, 32 bits) and double
 * (IEEE 754 double precision, 64 bits). Decimal literals default to double;
 * use the `f` suffix for float. `char` is a 1-byte integer holding an
 * ASCII (or other narrow encoding) code point — implicitly convertible to int.
 *
 * KEY POINTS:
 *  - 1.0     -> double (8 bytes)
 *  - 1.0f    -> float  (4 bytes)
 *  - 1.0L    -> long double (often 10/12/16 bytes, platform-dependent)
 *  - char ch = 'A';  static_cast<int>(ch) -> 65  (ASCII)
 *  - char + int promotes to int: 'a' + 1 == 98
 *  - For Unicode use char16_t / char32_t / wchar_t / std::u8string
 *  - <cfloat> exposes FLT_MIN/MAX, DBL_MIN/MAX, FLT_DIG, DBL_DIG, etc.
 *
 * SYNTAX:
 *   float f  = 1.5f;
 *   double d = 1.5;
 *   char  c  = 'X';            // 1 byte
 *   int   ic = c;               // 88 (implicit promotion)
 *   char  c2 = (char) 89;       // 'Y'
 *
 * DRY RUN:
 *  - char ch1=88, ch2='Y' -> "X Y"
 *  - 'a' + 1 -> 98 (int)
 *  - 'a' + 'b' -> 195
 */

#include <iostream>
#include <cfloat>
#include <climits>
#include <iomanip>

int main() {
    std::cout << std::left
              << std::setw(15) << "type" << std::setw(15) << "size (bytes)"
              << "  notes\n";
    std::cout << "----------------------------------------\n";
    std::cout << std::setw(15) << "char"        << std::setw(15) << sizeof(char)        << "  1 byte; ASCII code\n";
    std::cout << std::setw(15) << "wchar_t"     << std::setw(15) << sizeof(wchar_t)     << "  wide char (impl-defined)\n";
    std::cout << std::setw(15) << "char16_t"    << std::setw(15) << sizeof(char16_t)    << "  UTF-16 code unit\n";
    std::cout << std::setw(15) << "char32_t"    << std::setw(15) << sizeof(char32_t)    << "  UTF-32 code point\n";
    std::cout << std::setw(15) << "float"       << std::setw(15) << sizeof(float)       << "  IEEE 754 single (use 'f' suffix)\n";
    std::cout << std::setw(15) << "double"      << std::setw(15) << sizeof(double)      << "  IEEE 754 double (default for literals)\n";
    std::cout << std::setw(15) << "long double" << std::setw(15) << sizeof(long double) << "  extended precision\n";

    // Float vs double precision
    float  f = 10.4f;
    double d = 10.4;
    std::cout << std::fixed << std::setprecision(20)
              << "\nfloat  10.4f = " << f << '\n'
              << "double 10.4   = " << d << '\n';

    // Char arithmetic — exactly like Java
    char ch1 = 88;       // 'X'
    char ch2 = 'Y';
    std::cout << '\n' << ch1 << ' ' << ch2 << '\n';

    std::cout << "'a' + 1     = " << ('a' + 1) << '\n';        // 98 (int)
    std::cout << "'a' + 'b'  = " << ('a' + 'b') << '\n';       // 195
    std::cout << "static_cast<char>('a' + 1) = '"
              << static_cast<char>('a' + 1) << "'\n";          // 'b'

    return 0;
}

/*
 * NOTES:
 *  - C++'s `char` is 1 byte; Java's `char` is 2 bytes (Unicode UTF-16 code unit).
 *  - The 'f' suffix matters: `1.5` is double; `1.5f` is float.
 *  - For exact decimal arithmetic use a third-party library (no built-in equivalent
 *    of Python's `decimal.Decimal` in standard C++).
 *  - `static_cast<char>(int)` is the explicit narrowing conversion.
 */
